validation.go

```diff
@@ -306,7 +330,7 @@ func (v *validation) validate(vals []*validatorImpl, src peer.ID, msg *Message,
 	id := v.p.idGen.ID(msg)
 	if !v.p.markSeen(id) {
 		v.tracer.DuplicateMessage(msg)
-		return nil
+		return ValidationError{Reason: RejectValidationIgnoredDuplicate}
```

Comment:
I don't think we can do this. On surface, this appears like a sane thing to do but I suspect it's going to cause notable breakage downstream. The reason being that this will now error if the application attempts to publish a message that it already received from peers, which certainly happens in Ethereum when the CL successfully fetches blobs from the EL and broadcasts them. See the description here: https://github.com/libp2p/go-libp2p-pubsub/issues/610.

Reply:
ya, this is not so safe to change.

Reply:
Yeah, that's a good callout. Here's what we can do instead:

- Introduce a new duplicate error type (private)
- return this new duplicate error here
- callers check against this error and preserve all behavior
  - Topic.Publish no error, do nothing.
  - topic.AddToBatch no error, do not add the message to the batch.

Reply:
I made this change. Thinking about it more, it's actually reasonable to not return an error in `Publish`/`AddToBatch` for a duplicate case. Semantically we still published the message, even if it was in the past.

Reply:
Yep, I prefer restoring this and keeping this unchanged.
1. If this is deemed a duplicate because we observed an inbound copy, we already propagated it to peers in the past, so we can discard it silently.
2. If this is deemed a duplicate originating here (second Publish for same message ID), we should swallow it silently because our ability to sense duplicates is time-boxed anyway, causing a non-deterministic error which is hard to deal with. The max I would do here is log a WARN, to flag to the user that there may be some inconsistency in the app layer leading to duplicates. But even that is likely to be noisier than most users want tbh (it's simplier if they can push data to pubsub and trust that the layer will dedup where possible -- this being a feature and not an exception).